onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -216,6 +254,107 @@
   API_IMPL_END
 }
 
+bool IsHardwareAdapter(IDXCoreAdapter* adapter) {
+    bool is_hardware{ false };
+    THROW_IF_FAILED(adapter->GetProperty(
+        DXCoreAdapterProperty::IsHardware,
+        &is_hardware));
+    return is_hardware;
+}
+
+bool IsGPU(IDXCoreAdapter* compute_adapter) {
+    // Only considering hardware adapters
+    if (!IsHardwareAdapter(compute_adapter))
+        return false;
+    return compute_adapter->IsAttributeSupported(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GRAPHICS);
+}
+
+bool IsNPU(IDXCoreAdapter* compute_adapter) {
+    // Only considering hardware adapters
+    if (!IsHardwareAdapter(compute_adapter))
+        return false;
+    return !(compute_adapter->IsAttributeSupported(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GRAPHICS));
+}
+
+ORT_API_STATUS_IMPL(SessionOptionsAppendExecutionProvider_DML2, _In_ OrtSessionOptions* options, OrtDmlDeviceOptions* device_opts) {
+API_IMPL_BEGIN
+    OrtDmlPerformancePreference perf_pref = device_opts->perf_pref;
+    OrtDmlDeviceFilter dev_filter = device_opts->dev_filter;
+
+    // Create DXCore Adapter Factory
+    ComPtr<IDXCoreAdapterFactory> adapterFactory;
+    ORT_THROW_IF_FAILED(::DXCoreCreateAdapterFactory(adapterFactory.GetAddressOf()));
+
+    // Get a list of all the adapters that support compute
+    ComPtr<IDXCoreAdapterList> d3D12CoreComputeAdapters;
+    GUID attributes[]{ DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE };
+    ORT_THROW_IF_FAILED(
+        adapterFactory->CreateAdapterList(_countof(attributes),
+            attributes,
+            d3D12CoreComputeAdapters.GetAddressOf()));
+
+    const uint32_t count{ d3D12CoreComputeAdapters->GetAdapterCount() };
+    int compute_only_device_id = -1;
+
+    // Struct for holding each adapter
+    enum class DeviceType { GPU, NPU, BadDevice};
+    struct AdapterInfo {
+        ComPtr<IDXCoreAdapter> Adapter;
+        DeviceType Type; // GPU or NPU
+    };
+
+    // Returns an adapter type based on the user supplied device filter
+    std::function<DeviceType (IDXCoreAdapter*)> filterAdapterTypeQuery[4] = {
```

Comment:
>filterAdapterTypeQuery [](http://example.com/codeflow?start=48&length=22)

Alternatively extract this into its own method... probably cleaner, and make it return early based on a list:

DeviceType FilterAdapterTypeQuery(IDXCoreAdapter* adapter, OrtDmlDeviceFilter filter) {

if (filter & OrtDmlDeviceFilter::Gpu == OrtDmlDeviceFilter::Gpu) {
   return DeviceType::GPU;
}
if (filter & OrtDmlDeviceFilter::Npu == OrtDmlDeviceFilter::Npu) {
   return DeviceType::NPU;
}
return DeviceType::BadDevice;

}